airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergSchemaSynchronizer.kt

```diff
@@ -0,0 +1,138 @@
+     * @return The updated [Schema], after changes have been applied and committed.
+     */
+    fun applySchemaChanges(table: Table, incomingSchema: Schema): Schema {
+        val existingSchema = table.schema()
+        val diff = comparator.compareSchemas(incomingSchema, existingSchema)
+
+        if (!diff.hasChanges()) {
+            // If no differences, return the existing schema as-is.
+            return existingSchema
+        }
+
+        val update: UpdateSchema = table.updateSchema().allowIncompatibleChanges()
+
+        // 1) Remove columns that no longer exist in the incoming schema
+        diff.removedColumns.forEach { removedColumn -> update.deleteColumn(removedColumn) }
+
+        // 2) Update types => find a supertype for each changed column
+        diff.updatedDataTypes.forEach { columnName ->
+            val existingField =
+                existingSchema.findField(columnName)
+                    ?: error("Field \"$columnName\" not found in the existing schema!")
+            val incomingField =
+                incomingSchema.findField(columnName)
+                    ?: error("Field \"$columnName\" not found in the incoming schema!")
+
+            val superType: Type =
+                superTypeFinder.findSuperType(
+                    existingType = existingField.type(),
+                    incomingType = incomingField.type(),
+                    columnName = columnName
+                )
+            require(superType is PrimitiveType) {
+                "Currently only primitive type updates are supported. Attempted type: $superType"
+            }
+
+            // Update the column to the supertype
+            update.updateColumn(columnName, superType)
+        }
+
+        // 3) Mark columns newly optional
+        diff.newlyOptionalColumns.forEach { columnName -> update.makeColumnOptional(columnName) }
+
+        // 4) Add new columns, sorted by nesting depth (so that parents are created before children)
+        val sortedNewColumns =
+            diff.newColumns.sortedBy { it.count { char -> char == PARENT_CHILD_SEPARATOR } }
+
+        for (newColumnFqn in sortedNewColumns) {
+            val (parentPath, leafName) = splitIntoParentAndLeaf(newColumnFqn)
+
+            // Only 1-level nesting is supported
+            if (parentPath.count { it == PARENT_CHILD_SEPARATOR } > 0) {
+                throw IllegalArgumentException(
+                    "Adding nested columns more than 1 level deep is not supported: $newColumnFqn"
+                )
+            }
+
+            // Locate the appropriate incoming field
+            val incomingField =
+                if (parentPath.isEmpty()) {
+                    // Top-level column
+                    incomingSchema.findField(leafName)
+                        ?: error("Field \"$leafName\" not found in the incoming schema.")
+                } else {
+                    // 1-level nested column: "structFieldName~childField"
```

Comment:
seems fine for an initial release, but I think we should support fully recursive schema change before going full release - lots of API sources have deeply-nested structs

(I'm not actually sure if iceberg supports that, at least from just looking at the java api it provides - but we should check)